[dynamo][higher order op] Bugfixes to pass graph.lint

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102522
* #102450
* #102449
* #102462
* #102578
* __->__ #102448
* #102557
* #102547
* #102521


This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening, and graph.lint() was failing. I need `graph.lint()` in the followup PRs where I run an `Interpreter` on the subgraph to preserve the metadata information to AOT Autograd.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

anijain2305 committed:
[dynamo][higher order op] Bugfixes to pass graph.lint

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][higher order op] Bugfixes to pass graph.lint"



This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening, and graph.lint() was failing. I need `graph.lint()` in the followup PRs where I run an `Interpreter` on the subgraph to preserve the metadata information to AOT Autograd.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][higher order op] Bugfixes to pass graph.lint"



This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening, and graph.lint() was failing. I need `graph.lint()` in the followup PRs where I run an `Interpreter` on the subgraph to preserve the metadata information to AOT Autograd.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][higher order op] Bugfixes to pass graph.lint"



This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening, and graph.lint() was failing. I need `graph.lint()` in the followup PRs where I run an `Interpreter` on the subgraph to preserve the metadata information to AOT Autograd.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

zou3519 commented:
> This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening,

Do you have an example of what was actually going on?

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -782,26 +782,32 @@ def speculate_subgraph(
             for a in sub_args:
                 if a is None:
-                    a = ConstantVariable(None)
+                    new_arg = ConstantVariable(None)
                 assert not isinstance(
                     a, torch.Tensor
                 ), "Tensors should already be tracked?"
                 if isinstance(a, ConstantVariable):
                     proxy = tracer.create_graph_input("const")
+                    new_arg = a
```
What is going on here? if `a` is None, then we do not create an input for it in the subgraph?

Also, is it possible to add a test?

zou3519 commented:
> > This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening,
> 
> Do you have an example of what was actually going on?

Nvm, I see what was going on

anijain2305 committed:
Update on "[dynamo][higher order op] Bugfixes to pass graph.lint"



This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening, and graph.lint() was failing. I need `graph.lint()` in the followup PRs where I run an `Interpreter` on the subgraph to preserve the metadata information to AOT Autograd.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -782,26 +782,32 @@ def speculate_subgraph(
+                    new_arg = a
                 elif isinstance(a, (TensorVariable, AutogradFunctionContextVariable)):
-                    tracer.create_graph_input(a.as_proxy().node.name)
+                    new_proxy = tracer.create_graph_input(a.as_proxy().node.name)
+                    example_value = a.as_proxy().node.meta["example_value"]
+                    new_arg = wrap_fx_proxy(
+                        tx=tx, proxy=new_proxy, example_value=example_value
+                    )
```
This looks like it's causing one of the autograd.Function test cases to fail? `ReproTests.test_hf_xsoftmax_training`

anijain2305 committed:
Update on "[dynamo][higher order op] Bugfixes to pass graph.lint"



This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening, and graph.lint() was failing. I need `graph.lint()` in the followup PRs where I run an `Interpreter` on the subgraph to preserve the metadata information to AOT Autograd.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][higher order op] Bugfixes to pass graph.lint"



This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening, and graph.lint() was failing. I need `graph.lint()` in the followup PRs where I run an `Interpreter` on the subgraph to preserve the metadata information to AOT Autograd.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][higher order op] Bugfixes to pass graph.lint"



This PR ensures that the subgraphs use the newly created placeholder for the primary inputs and free variables. Earlier, this was not happening, and graph.lint() was failing. I need `graph.lint()` in the followup PRs where I run an `Interpreter` on the subgraph to preserve the metadata information to AOT Autograd.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -789,20 +790,30 @@ def speculate_subgraph(
+                    # future, we can clean this up.
+                    tracer.create_graph_input("const")
+                    # Ensures that we recompile when the constant value changes
+                    a.add_guard(GuardBuilder.CONSTANT_MATCH)
```
One thing I am curious about is: why doesn't a guard on the constant get added when dynamo sees it being used?

In the following test case,
```py
        x = torch.randn(3, 3)
        y = 4.0

        def fn(x, y, z):
            if z:
                return x + y
            return x * y

        def f(x, y, z):
            return wrap(fn, x, y, z)

        args = (x, 4.0, None)
        opt_f = torch.compile(f, fullgraph=True, backend=CompileCounter())
```
I would have naively expected dynamo to add a guard when it sees `x * y` and notices that `y` is a constant

anijain2305 replied:
That is a good question! 
Yeah, I don't know. Let's revisit this. I am going to note this down in tracker.

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge
